pkg/dio/multi_test.go

```diff
@@ -12,365 +12,278 @@
+	if err != nil {
+		t.Errorf("write: %v", err)
+	}
+	if n != len(p) {
+		t.Errorf("short write: %d != %d", n, len(p))
+	}
 }
 
-func TestMultiWrite(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+func read(t *testing.T, mreader DynamicMultiReader, limit int) []byte {
+	total := 0
 
-	mwriter := MultiWriter(pipeAW, pipeBW)
+	var buf = make([]byte, 32*1024)
+	for {
+		n, err := mreader.Read(buf)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			t.Errorf("read: %v", err)
+		}
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+		total += n
+		if total >= limit {
+			break
+		}
+	}
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+	return buf[:limit]
+}
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+func control(t *testing.T, buffers []*bytes.Buffer, p []byte) {
+	for i := range buffers {
+		buffer := buffers[i]
+		if buffer.String() != string(p) {
+			t.Errorf("Expected: %q, actual: %q", string(p), buffer.String())
+		}
 	}
+}
 
+func equal(t *testing.T, mreader DynamicMultiReader, str string, count int) {
+	expected := strings.Repeat(str, count)
+	// read from multi reader
+	buffer := string(read(t, mreader, len(expected)))
 	// compare the data
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
-	}
-
-	if bufB.String() != data {
-		t.Errorf("B: expected: %s, actual: %s", data, bufB.String())
-		return
+	if buffer != expected {
+		t.Errorf("Expected: %q, actual: %q", expected, buffer)
 	}
 }
 
-func TestWriteAdd(t *testing.T) {
-	pipeAR, pipeAW := io.Pipe()
-	pipeBR, pipeBW := io.Pipe()
+// TestMultiWrite creates multi writers and writes to them, then removes some of them and writes again
+func TestMultiWrite(t *testing.T) {
+	var writers []io.Writer
+	var buffers []*bytes.Buffer
 
-	mwriter := MultiWriter(pipeAW)
+	// create & initialize writers and buffers
+	for i := 0; i < count; i++ {
+		var buffer bytes.Buffer
 
-	var bufA bytes.Buffer
-	var bufB bytes.Buffer
+		reader, writer := io.Pipe()
 
-	// set up a copy so we don't block writes
-	go io.Copy(&bufA, pipeAR)
-	go io.Copy(&bufB, pipeBR)
+		writers = append(writers, writer)
+		buffers = append(buffers, &buffer)
 
-	// send the test string
-	data := "verify base multiwriter function"
-	_, err := mwriter.Write([]byte(data))
-	if err != nil {
-		t.Error(err)
-		return
+		// set up a goroutine so we don't block writes
+		go io.Copy(&buffer, reader)
 	}
 
-	// compare the data - shouldn't be present
-	if bufA.String() != data {
-		t.Errorf("A: expected: %s, actual: %s", data, bufA.String())
-		return
-	}
+	// create the multi writer
+	mwriter := MultiWriter(writers...)
 
-	if bufB.String() != "" {
-		t.Errorf("B: expected: %s, actual: %s", "", bufB.String())
-		return
-	}
+	write(t, mwriter, []byte(base))
+	control(t, buffers, []byte(base))
```

Comment:
how do we ensure that all of the io.Copy's have finished by the time we compare buffers?